Modules/Core/TestKernel/include/itkTestDriverIncludeBuiltInFactories.h

```diff
@@ -44,6 +44,28 @@
 #include "itkOFFMeshIOFactory.h"
 #include "itkVTKPolyDataMeshIOFactory.h"
 
+// FFT
+#include "itkFFTImageFilterFactory.h"
+#include "itkVnlComplexToComplex1DFFTImageFilter.h"
+#include "itkVnlComplexToComplexFFTImageFilter.h"
+#include "itkVnlForward1DFFTImageFilter.h"
+#include "itkVnlForwardFFTImageFilter.h"
+#include "itkVnlHalfHermitianToRealInverseFFTImageFilter.h"
+#include "itkVnlInverse1DFFTImageFilter.h"
+#include "itkVnlInverseFFTImageFilter.h"
+#include "itkVnlRealToHalfHermitianForwardFFTImageFilter.h"
+
+#if defined(ITK_USE_FFTWD) || defined(ITK_USE_FFTWF)
+#  include "itkFFTWComplexToComplex1DFFTImageFilter.h"
+#  include "itkFFTWComplexToComplexFFTImageFilter.h"
+#  include "itkFFTWForward1DFFTImageFilter.h"
+#  include "itkFFTWForwardFFTImageFilter.h"
+#  include "itkFFTWHalfHermitianToRealInverseFFTImageFilter.h"
+#  include "itkFFTWInverse1DFFTImageFilter.h"
+#  include "itkFFTWInverseFFTImageFilter.h"
+#  include "itkFFTWRealToHalfHermitianForwardFFTImageFilter.h"
+#endif
```

Comment:
For test purposes, we only need the VNL implementations -- this can be removed.

Reply:
Thanks @thewtex !

When available, FFTW implementations are registered by default alongside VNL implementations when `UseITK.cmake` is included in a project consuming ITK (see [https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/Filtering/FFT/itk-module.cmake#L11](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/Filtering/FFT/itk-module.cmake#L11)). For the sake of consistency, isn't it better for the TestKernel act in the same way?

To me it seems that consistency is preserved if the TestKernel and any consuming project follow the same rules for default registrations:
- If FFTW is available, register FFTW factories;
- Always register VNL factories

We should also note that the principle purpose of these changes is to reduce boilerplate for FFT factory registrations in ITK tests. We do have tests that expect FFTW factory overrides if FFTW is available, such as [itkFFT1DImageFilterTest.cxx](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/Filtering/FFT/test/itkFFT1DImageFilterTest.cxx#L109). Removing FFTW factories from default registration in the TestKernel would mean reintroducing that boilerplate or altering test coverage.

Reply:
I agree with Tom, it is better to have this here, once, than many times over in different tests. FFTW is not built by default, so VNL backend will be exercised much more often.

Reply:
This patch provide consistency with the previous Vnl and FFTW testing infrastructure, but that is not what I meant by consistency.

In unit testing, we want

- Tests that check "one thing"
- Consistency in what is tested

> When available, FFTW implementations are registered by default alongside VNL implementations when UseITK.cmake is included in a project consuming ITK (see https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/Filtering/FFT/itk-module.cmake#L11). For the sake of consistency, isn't it better for the TestKernel act in the same way?

No, we do not want the same behavior for application projects and internal ITK unit testing. For deployed applications, we want the application to be able to load the best FFT backend available at runtime. For unit testing in the TestKernel, we want consistent behavior to occur in the unit tests.

Tests that require an FFT-based filter should be able to run the tests without additional instrumentation, e.g. registering and FFT backend. However, they should get the same backend from the TestKernel.  If the goal is to also test functionality with additional FFT backends, then those backend classes should be instantiated explicitly and explicitly tested in the tests. To test additional backends in unit tests, the factory mechanism should not be used. FFTW is not the only additional backend. There is VkFFT and hopefully more in the future. By explicitly using the FFT backend classes in the unit tests, the tested behavior is consistent and we get coverage. Registering FFT-backends in this way for FFTW cannot be completed with other FFT backends, which may be added in remote module builds that occur after ITK has been built. Also, the factory does not yet have a priority mechanism, and the FFT implementation may depend on the registration or other mechanism. It is better to be explicit with the registered methods.

Reply:
Thanks for the explanation; I've pushed a WIP commit for removing default FFTW registrations and will squash into the relevant TestKernel commit if all tests pass.